scripts/core/sharedContext.yml

```diff
@@ -0,0 +1,88 @@
+#
+# Copyright (C) 2023 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+#
+# See YaML.md for syntax definition
+#
+--- #--------------------------------------------------------------------------
+type: header
+desc: "Intel $OneApi Level-Zero Extension for shared contexts."
+version: "1.6"
+--- #--------------------------------------------------------------------------
+type: macro
+desc: "Shared Contexts Extension Name"
+version: "1.6"
+name: $X_SHARED_CONTEXTS_EXP_NAME
+value: '"$X_experimental_shared_contexts"'
+--- #--------------------------------------------------------------------------
+type: enum
+desc: "Shared Contexts Extension Version(s)"
+version: "1.6"
+name: $x_shared_contexts_exp_version_t
+etors:
+    - name: "1_0"
+      value: "$X_MAKE_VERSION( 1, 0 )"
+      desc: "version 1.0"
+--- #--------------------------------------------------------------------------
+type: function
+desc: "Defines a shared context for a set of context handles."
+version: "1.6"
+class: $xDriver
+name: GetSharedContextExp
+details:
+    -  |
+       If a shared context has been previously created for the contexts passed in the array,
+       and the shared-context is the same for all, then that one is returned.
+    -  |
+       If none of the contexts in the array belong to a shared context, then a shared context is created
+       encompassing all the contexts in the array.
+    -  |
+       If some of the contexts in the array are not yet associated with any shared context, and the rest
+       belong to a single shared context, then all are added to the same shared context.
+    -  |
+       If some of the contexts in the array already belong to different shared contexts,
+       then nullptr is returned.
+    -  |
+       If only one context is passed, then its shared context is returned if it already exists.
+       If not, then a new shared context is created and returned. Additional contexts
+       can then later be added to this one using $xDriverAddContextToSharedExp.
+    - "If only one context is passed, and that context is a shared context, then the same is returned."
+    - "The application may call this function from simultaneous threads with the same context handle."
+    - "The implementation of this function should be lock-free."
+params:
+    - type: $x_driver_handle_t
+      name: hDriver
+      desc: "[in] handle of driver object"
+    - type: uint32_t
+      name: count
+      desc: "[in] number of context passed in hContexts array"
+    - type: "$x_context_handle_t*"
+      name: phContexts
+      desc: "[in][range(0, count)] array of contexts to create the new shared-context with"
+    - type: "$x_context_handle_t*"
+      name: phSharedContext
+      desc: "[out] pointer to the shared context object"
+--- #--------------------------------------------------------------------------
+type: function
+desc: "Adds a context to an existing shared context."
+version: "1.6"
+class: $xDriver
+name: AddContextToSharedExp
```

Comment:
how is context removed from shared context?

Reply:
Does the API need a mechanism for this, or can the driver/runtime do it during context teardown?

Reply:
@MichalMrozek thanks. IWe could add zeDriverRemoveContextFromShared.

Reply:
We need to be careful with this extension to not relax the current boundaries too much. One problem that I see is that it allows resources to be avaiblle automatically outside of the context where they were allocated.

Reply:
@MichalMrozek . Correct, that's the main idea. What happens is that we could have multi-library applications (MPI + SYCL + MKL), where USM allocations are created by one library and then their pointers exchanged with others (like SYCL allocating the memory, and the MPI using it in accelerator collectives). So the only way they could use those pointers coming from other library is having a way to share the resource with their own context, which is where these APIs come into play, allowing users to explicitly request such sharing.

This is needed to start enforcing per-context resource isolation in the L0 GPU driver. We need to either add this or abandon for completely the idea of isolating resources per context.

Reply:
Context is pretty fundamental approach in the API that provides encapsulation.
Enabling dynamic sharing is very challenging to deliver, basically all resources that you allocated prior to creation of shared context, now need to be re-evaluated on every subsequent submit, which would substantially add to overheads.

There are also other ways to solve this:

1. use import / export

With export / import notion you can still export resource from one context and import it to another.
It also clearly defines lifecycle and it connected to the resource.

2. add default context

Another way to solve this is to allow default context per driver instance. This would be simple getDefaultContext API that returns default context, SYCL uses it for all allocations.

Reply:
thanks @MichalMrozek . 

The idea of the default context is good, but I think that @wdamon-intel proposed this approach since it is more explicit. With default context, you have a default context belonging to all contexts, but with these APIs, user is doing a more explicitly opt-in for this feature. @wdamon-intel can elaborate further.

I like the idea of the import/export. Basically, instead of creating a shared-context and having all resources being shared, with import/export the sharing is more fine granular and specific to a given resources, instead of all. Also, we avoid the issues with shared-contexts that we are discussing here.

We could have the following:

```cpp
zeContextExportResource(hContext, *void handle)
the USM pointer or handle passed as void is marked by current library as exportable,
meaning that it can be imported by other contexts.

zeContextImportResource(hContext, *void handle)
the USM pointer or handle passed as void is imported into the context to be used with
any resource created with it.
```

@MichalMrozek is that more or less what you were thinking?

@wdamon-intel what do you think?

Reply:
The basic idea was to create a mechanism by which libraries/instances (within the same process address space) could share resources while retaining isolation/separation as there is today. The "shared context" concept comes from OpenGL. There, a context is created with another context handle and they belong to the same "group" as far as the implementation is concerned. **There is no such thing as an explicit "shared context object"**. The implementation maintains the resource isolation between contexts, but contexts part of a share-group may directly access (most of) one-another's objects. The only way a context joins a group is at creation time, and the only way it leaves a group is by being destroyed.

To keep it simple, I think we could:

1. Abandon this proposal and just leverage import/export. Or,
2. Simplify this proposal and adopt some of the OpenGL share group rules discussed above.